gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAChallengeEndpoint.java

```diff
@@ -325,11 +358,34 @@ private void redirectToAuthorize(RoutingContext routingContext, Client client) {
         }
     }
 
-    private void verify(FactorProvider factorProvider, FactorContext factorContext, Handler<AsyncResult<Void>> handler) {
+    private void verify(FactorProvider factorProvider, FactorContext factorContext, Optional<VerifyAttempt> verifyAttempt,
+                        RoutingContext routingContext, Handler<AsyncResult<Void>> handler) {
         factorProvider.verify(factorContext)
                 .subscribe(
-                        () -> handler.handle(Future.succeededFuture()),
-                        error -> handler.handle(Future.failedFuture(error)));
+                        () -> {
+                            if (verifyAttempt.isPresent()) {
+                                verifyAttemptService.delete(verifyAttempt.get().getId()).subscribe(
+                                        () -> handler.handle(Future.succeededFuture()),
+                                        error -> {
+                                            logger.error("Could not delete verify attempt", error);
+                                            routingContext.fail(401);
+                                        }
+                                );
+                            } else {
+                                handler.handle(Future.succeededFuture());
+                            }
+                        },
+                        error -> {
+                            final EnrolledFactor enrolledFactor = (EnrolledFactor) factorContext.getData().get(FactorContext.KEY_ENROLLED_FACTOR);
+                            verifyAttemptService.verificationFailed(factorContext.getUser().getId(), enrolledFactor.getFactorId(),
+                                    factorContext.getClient(), domain, verifyAttempt).subscribe(
+                                    () -> handler.handle(Future.failedFuture(error)),
+                                    verificationFailedError -> {
+                                        logger.error("Could not updated verification failed status", verificationFailedError);
+                                        routingContext.fail(401);
```

Comment:
Log the error is OK but IMHO we should respond to the client with the validation error like we are doing at line 382

`                                    () -> handler.handle(Future.failedFuture(error))`

wdyt ?

Reply:
done